frame/nft-fractionalization/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+		/// - `beneficiary`: The account that will receive the newly created asset.
+		/// - `fractions`: The amount to be minted of the newly created asset.
+		///
+		/// Emits `NftFractionalized` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::fractionalize())]
+		pub fn fractionalize(
+			origin: OriginFor<T>,
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+			fractions: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			let nft_owner =
+				T::Nfts::owner(&nft_collection_id, &nft_id).ok_or(Error::<T>::NftNotFound)?;
+			ensure!(nft_owner == who, Error::<T>::NoPermission);
+
+			let pallet_account = Self::get_pallet_account();
+			let deposit = T::Deposit::get();
+			T::Currency::reserve(&nft_owner, deposit)?;
+			Self::do_lock_nft(nft_collection_id, nft_id)?;
+			Self::do_create_asset(asset_id, pallet_account.clone())?;
+			Self::do_mint_asset(asset_id, &beneficiary, fractions)?;
+			Self::do_set_metadata(asset_id, &who, &pallet_account, &nft_collection_id, &nft_id)?;
+
+			NftToAsset::<T>::insert(
+				(nft_collection_id, nft_id),
+				Details { asset: asset_id, fractions, asset_creator: nft_owner, deposit },
+			);
+
+			Self::deposit_event(Event::NftFractionalized {
+				nft_collection: nft_collection_id,
+				nft: nft_id,
+				fractions,
+				asset: asset_id,
+				beneficiary,
+			});
+
+			Ok(())
+		}
+
+		/// Burn the whole amount of the asset and return back the locked NFT.
+		///
+		/// The dispatch origin for this call must be Signed.
+		///
+		/// `Deposit` funds will be returned to `asset_creator`.
+		///
+		/// - `nft_collection_id`: The ID used to identify the collection of the NFT.
+		/// Is used within the context of `pallet_nfts`.
+		/// - `nft_id`: The ID used to identify the NFT within the given collection.
+		/// Is used within the context of `pallet_nfts`.
+		/// - `asset_id`: The ID of the asset being returned and destroyed. Must match
+		/// the original ID of the created asset, corresponding to the NFT.
+		/// Is used within the context of `pallet_assets`.
+		/// - `beneficiary`: The account that will receive the unified NFT.
+		///
+		/// Emits `NftUnified` event when successful.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::unify())]
+		pub fn unify(
+			origin: OriginFor<T>,
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			NftToAsset::<T>::try_mutate_exists((nft_collection_id, nft_id), |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::DataNotFound)?;
+				ensure!(details.asset == asset_id, Error::<T>::DataNotFound);
```

Comment:
Could the first error be called `Error::<T>::NftNotFractionalised` and the second be `Error::<T>::IncorrectAssetId` ? It really helps to know exactly what's gone wrong.

Reply:
Good point